Assets/BossRoom/Scripts/Client/UI/Lobby/JoinLobbyUI.cs

```diff
@@ -0,0 +1,160 @@
+using System.Collections.Generic;
+using BossRoom.Scripts.Shared.Infrastructure;
+using BossRoom.Scripts.Shared.Net.UnityServices.Lobbies;
+using UnityEngine;
+using UnityEngine.UI;
+
+namespace BossRoom.Scripts.Client.UI
+{
+    /// <summary>
+    /// Handles the list of LobbyButtons and ensures it stays synchronized with the lobby list from the service.
+    /// </summary>
+    public class JoinLobbyUI : ObserverBehaviour<LobbyServiceData>
+    {
+        private UIFactory m_UIFactory;
+        private LobbyUIManager m_LobbyUIManager;
+        private LobbyServiceData m_LobbyServiceData;
+
+        [SerializeField] private LobbyPanelUI m_LobbyPanelPrototype;
+
+        //
+        [SerializeField]
+        InputField m_JoinCodeField;
+
+        /// <summary>
+        /// Key: Lobby ID, Value Lobby UI
+        /// </summary>
+        Dictionary<string, LobbyPanelUI> m_LobbyButtons = new Dictionary<string, LobbyPanelUI>();
+        Dictionary<string, LocalLobby> m_LocalLobby = new Dictionary<string, LocalLobby>();
+
+        /// <summary>Contains some amount of information used to join an existing lobby.</summary>
+        LocalLobby.LobbyData m_LocalLobbySelected;
+
+
+        [Inject]
+        private void InjectDependencies(UIFactory uiFactory, LobbyUIManager lobbyUIManager, LobbyServiceData lobbyServiceData)
+        {
+            m_UIFactory = uiFactory;
+            m_LobbyUIManager = lobbyUIManager;
+            m_LobbyServiceData = lobbyServiceData;
+        }
+
+        private void Awake()
+        {
+            m_LobbyPanelPrototype.gameObject.SetActive(false);
+        }
+
+        public void LobbyPanelSelected(LocalLobby lobby)
+        {
+            m_LocalLobbySelected = lobby.Data;
+        }
+
+        public void OnLobbyCodeInputFieldChanged(string newCode)
+        {
+            if (!string.IsNullOrEmpty(newCode))
+                m_LocalLobbySelected = new LocalLobby.LobbyData(newCode.ToUpper());
+        }
+
+        public void OnJoinButtonPressed()
+        {
+            m_LobbyUIManager.JoinLobbyRequest(m_LocalLobbySelected);
+            m_LocalLobbySelected = default;
+        }
+
+        public void OnRefresh()
+        {
+            m_LobbyUIManager.QueryLobbiesRequest();
+        }
+
+        protected override void UpdateObserver(LobbyServiceData observed)
+        {
+            ///Check for new entries, We take CurrentLobbies as the source of truth
+            var previousKeys = new List<string>(m_LobbyButtons.Keys);
+
+            foreach (var codeLobby in observed.CurrentLobbies)
+            {
+                var lobbyCodeKey = codeLobby.Key;
+                var lobbyData = codeLobby.Value;
+
+                if (!m_LobbyButtons.ContainsKey(lobbyCodeKey))
```

Comment:
Each Lobby has an ID which is guaranteed to be unique and won't change.  It's _possible_ for the lobby code to change so it shouldn't be used as a key.

Reply:
Okay, maybe this is just a naming problem?  Lobbies have a code (short alphanumeric intended to be shared and manually typed in to join a room), and an ID (unique GUIDy thing).  I believe that m_LobbyButtons is keyed by lobbyId and this is just named wrong.  Correct me if it's not.